File: source/bxgenbb_help/src/particle.f
c particle.f 
c
c This file was extracted from the 'decay0' program by V.I. Tretyak
c
c Copyright 1995-2011 V.I. Tretyak
c
c This program is free software; you can redistribute it and/or modify
c it under the terms of the GNU General Public License as published by
c the Free Software Foundation; either version 3 of the License, or (at
c your option) any later version.
c 
c This program is distributed in the hope that it will be useful, but
c WITHOUT ANY WARRANTY; without even the implied warranty of
c MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
c General Public License for more details.
c 
c You should have received a copy of the GNU General Public License
c along with this program; if not, write to the Free Software
c Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
c 
c
	subroutine particle(np,E1,E2,teta1,teta2,phi1,phi2,tclev,
     +                      thlev,tdlev)
c Generation of isotropical emission of particle in the range of
c energies and angles.
c Call  : common/genevent/tevst,npfull,npgeant(100),pmoment(3,100),ptime(100)
c         call particle(np,E1,E2,teta1,teta2,phi1,phi2,tclev,thlev,tdlev)
c Input : np          - GEANT number for particle identification:
c                        1 - gamma         2 - positron     3 - electron
c                        4 - neutrino      5 - muon+        6 - muon-
c                        7 - pion0         8 - pion+        9 - pion-
c                       10 - kaon0 long   11 - kaon+       12 - kaon-
c                       13 - neutron      14 - proton      15 - antiproton
c                       16 - kaon0 short  17 - eta         18 - lambda
c                       19 - sigma+       20 - sigma0      21 - sigma-
c                       22 - xi0          23 - xi-         24 - omega
c                       25 - antineutron  26 - antilambda  27 - antisigma-
c                       28 - antisigma0   29 - antisigma+  30 - antixi0
c                       31 - antixi+      32 - antiomega+  45 - deuteron
c                       46 - tritium      47 - alpha       48 - geantino
c                       49 - He3          50 - Cerenkov;
c         E1,E2       - range of kinetic energy of particle (MeV);
c         teta1,teta2 - range of teta angle (radians);
c         phi1,phi2   - range of phi  angle (radians);
c         tclev       - time of creation of level from which particle will be
c                       emitted (sec);
c         thlev       - level halflife (sec).
c Output: tdlev               - time of decay of level (sec);
c         tevst               - time of event's start (sec);
c         npfull              - current number of particle in event;
c         npgeant(npfull)     - GEANT number for particle identification
c                               (=np of input);
c         pmoment(1-3,npfull) - x,y,z components of particle momentum (MeV);
c         ptime(npfull)       - time shift from previous time to calculate
c                               when particle was emitted (sec).
c VIT, 15.10.1995.
	common/genevent/tevst,npfull,npgeant(100),pmoment(3,100),
     +                  ptime(100)
	common/const/pi,emass,datamass(50)
	npfull=npfull+1
	if(npfull.gt.100) print *,
     +    'PARTICLE: in event more than 100 particles ',npfull
c	if(np.lt.1.or.np.gt.50.or.(np.gt.32.and.np.lt.45)) print *,
c    1     'PARTICLE: unknown particle number ',np
	npgeant(npfull)=np
	pmass=datamass(np)
	phi=phi1+(phi2-phi1)*rnd1(d)
	ctet1=1.
	if(teta1.ne.0.) ctet1=cos(teta1)
	ctet2=-1.
	if(teta2.ne.pi) ctet2=cos(teta2)
	ctet=ctet1+(ctet2-ctet1)*rnd1(d)
	stet=sqrt(1.-ctet*ctet)
	E=E1
	if(E1.ne.E2) E=E1+(E2-E1)*rnd1(d)
	p=sqrt(E*(E+2.*pmass))
	pmoment(1,npfull)=p*stet*cos(phi)
	pmoment(2,npfull)=p*stet*sin(phi)
	pmoment(3,npfull)=p*ctet
	tdlev=tclev
	if(thlev.gt.0.) tdlev=tclev-thlev/alog(2.)*alog(rnd1(d))
	ptime(npfull)=tdlev
	return
	end
c
c end of particle.f
c
c Local Variables: --
c mode: fortran --
c End: --

File: source/bxgenbb_help/src/decay0/particle.cc

#include <cmath>
#include <limits>
#include <stdexcept>

#include <datatools/units.h>
#include <mygsl/rng.h>
#include <genbb_help/decay0/particle.h>
#include <genbb_help/primary_event.h>

namespace genbb {
  namespace decay0 {

    /// Return the name of the particle with given ID (MeV)
    const std::string decay0_geant3_particle_name(int np)
    {
     switch(np) {
        case GAMMA: return std::string("gamma");
        case POSITRON: return std::string("positron");
        case ELECTRON: return std::string("electron");
        case NEUTRINO: return std::string("neutrino");
        case MUONPLUS: return std::string("muon+");
        case MUONMINUS: return std::string("muon-");
        case PION0: return std::string("pion0");
        case PIONPLUS: return std::string("pion+");
        case PIONMINUS: return std::string("pion-");
        case KAON0LONG: return std::string("kaon0 long");
        case KAONPLUS: return std::string("kaon+");
        case KAONMINUS: return std::string("kaon-");
        case NEUTRON: return std::string("neutron");
        case PROTON: return std::string("proton");
        case ANTIPROTON: return std::string("antiproton");
        case KAON0SHORT: return std::string("kaon0 short");
        case ETA: return std::string("eta");
        case LAMBDA: return std::string("lambda");
        case SIGMAPLUS: return std::string("sigma+");
        case SIGMA0: return std::string("sigma0");
        case SIGMAMINUS: return std::string("sigma-");
        case XI0: return std::string("xi0");
        case XIMINUS: return std::string("xi-");
        case OMEGA: return std::string("omega");
        case ANTINEUTRON: return std::string("antineutron");
        case ANTILAMBDA: return std::string("antilambda");
        case ANTISIGMAMINUS: return std::string("antisigma-");
        case ANTISIGMA0: return std::string("antisigma0");
        case ANTISIGMAPLUS: return std::string("antisigma+");
        case ANTIXI0: return std::string("antixi0");
        case ANTIXIPLUS: return std::string("antixi+");
        case ANTIOMEGAPLUS: return std::string("antiomega+");
        case DEUTERON: return std::string("deuteron");
        case TRITIUM: return std::string("tritium");
        case ALPHA: return std::string("alpha");
        case GEANTINO: return std::string("geantino");
        case HE3: return std::string("He3");
        case CERENKOV: return std::string("Cerenkov");
        }
     return std::string("");
    }

    double decay0_particle_mass(int np)
    {
      switch(np) {
        case GAMMA: return 0.;
        case POSITRON: return 0.51099906;
        case ELECTRON: return 0.51099906;
        case NEUTRINO: return 0.;
        case MUONPLUS: return 105.658389;
        case MUONMINUS: return 105.658389;
        case PION0: return 134.9764;
        case PIONPLUS: return 139.5700;
        case PIONMINUS: return 139.5700;
        case KAON0LONG: return 497.672;
        case KAONPLUS: return 493.677;
        case KAONMINUS: return 493.677;
        case NEUTRON: return 939.56563;
        case PROTON: return 938.27231;
        case ANTIPROTON: return 938.27231;
        case KAON0SHORT: return 497.672;
        case ETA: return 547.45;
        case LAMBDA: return 1115.684;
        case SIGMAPLUS: return 1189.37;
        case SIGMA0: return 1192.55;
        case SIGMAMINUS: return 1197.436;
        case XI0: return 1314.9;
        case XIMINUS: return 1321.32;
        case OMEGA: return 1672.45;
        case ANTINEUTRON: return 939.56563;
        case ANTILAMBDA: return 1115.684;
        case ANTISIGMAMINUS: return 1189.37;
        case ANTISIGMA0: return 1192.55;
        case ANTISIGMAPLUS: return 1197.436;
        case ANTIXI0: return 1314.9;
        case ANTIXIPLUS: return 1321.32;
        case ANTIOMEGAPLUS: return 1672.45;
        case DEUTERON: return 1875.613;
        case TRITIUM: return 2809.25;
        case ALPHA: return 3727.417;
        case GEANTINO: return 0.;
        case HE3: return 2809.23;
        case CERENKOV: return 0.;
        }
      return std::numeric_limits<double>::quiet_NaN();
    }

    double decay0_emass()
    {
      return decay0_particle_mass(ELECTRON);
    }

    void decay0_particle(mygsl::rng & prng_, primary_event & event_,
                         int np, double E1, double E2,
                         double teta1, double teta2,
                         double phi1, double phi2,
                         double tclev, double thlev, double & tdlev)
    {
      double last_time = 0.0 * CLHEP::second;
      if (event_.get_particles ().size() > 0) {
        const primary_particle & last_part = event_.get_particles ().back();
        last_time = last_part.get_time();
      }

      primary_particle part;
      double pmass = decay0_particle_mass (np);
      if (isnan(pmass)) {
        std::ostringstream message;
        message << "genbb::decay0::decay0_particle: "
                << "Unknown mass for particle with ID=" << np << " !";
        throw std::logic_error(message.str());
      }
      double phi=phi1+(phi2-phi1)*prng_();
      double ctet1=1.;
      if (teta1 != 0.) ctet1=std::cos(teta1);
      double ctet2=-1.;
      if (teta2 != M_PI) ctet2=std::cos(teta2);
      double ctet=ctet1+(ctet2-ctet1)*prng_();
      double stet=std::sqrt(1.-ctet*ctet);
      double E=E1;
      if (E1 != E2) E=E1+(E2-E1)*prng_();
      // --- randomly blur the particle energy
      // double dE=0.000025;
      // E=E+dE*(-1+2*prng_());
      double p=std::sqrt(E*(E+2.*pmass));
      geomtools::vector_3d momentum(p*stet*std::cos(phi),
                                    p*stet*std::sin(phi),
                                    p*ctet);
      geomtools::vector_3d vertex;
      geomtools::invalidate(vertex);

      tdlev=tclev;
      if (thlev > 0.) tdlev = tclev - thlev / std::log(2.) * std::log(prng_());

      part.set_type(np);
      part.set_time (last_time + tdlev * CLHEP::second);
      part.set_momentum (momentum * CLHEP::MeV);
      part.set_vertex (vertex);
      event_.grab_particles ().push_back(part);

      return;
    }

  } // end of namespace decay0
} // end of namespace genbb
